src/lib/profiles/software-update/SoftwareUpdateProfile.cpp

```diff
@@ -28,460 +28,575 @@
+ *
+ * @param[in] aLength   An 8-bit value for the length of the list.  Must be less that the number of enums in @ref UpdateSchemes.
+ * @param[in] aList  A pointer to an array of @ref UpdateSchemes values.  May be NULL only if aLength is 0.
+ *
+ * @retval WEAVE_NO_ERROR  On success.
+ * @retval WEAVE_ERROR_INVALID_LIST_LENGTH If the length is too long
  */
-WEAVE_ERROR UpdateSchemeList::init(uint8_t aLength, uint8_t *aList)
+WEAVE_ERROR UpdateSchemeList::init(uint8_t aLength, uint8_t * aList)
 {
-  if (aLength>4) return WEAVE_ERROR_INVALID_LIST_LENGTH;
-  theLength=aLength;
-  for (int i=0; i<theLength; i++) theList[i]=aList[i];
-  return WEAVE_NO_ERROR;
+    if (aLength > 4)
+        return WEAVE_ERROR_INVALID_LIST_LENGTH;
+    theLength = aLength;
+    for (int i = 0; i < theLength; i++)
+        theList[i] = aList[i];
+    return WEAVE_NO_ERROR;
 }
-/*
- * parameters: MessageIterator &i,
- * returns: error/status
+
+/**
+ * Serialize the object to the provided MessageIterator
+ *
+ * @param[in] i An iterator over the message being packed
+ *
+ * @retval WEAVE_NO_ERROR  On success.
+ * @retval WEAVE_ERROR_BUFFER_TOO_SMALL If the list is too long to fit in the message.
  */
-WEAVE_ERROR UpdateSchemeList::pack(MessageIterator &i)
+WEAVE_ERROR UpdateSchemeList::pack(MessageIterator & i)
 {
-  if (i.hasRoom(theLength+1)) {
-    i.writeByte(theLength);
-    for (int j=0; j<theLength; j++) i.writeByte(theList[j]);
-    return WEAVE_NO_ERROR;
-  }
-  else
-    return WEAVE_ERROR_INVALID_LIST_LENGTH;
+    if (i.hasRoom(theLength + 1))
+    {
+        i.writeByte(theLength);
+        for (int j = 0; j < theLength; j++)
+            i.writeByte(theList[j]);
+        return WEAVE_NO_ERROR;
+    }
+    else
+        return WEAVE_ERROR_BUFFER_TOO_SMALL;
 }
-/*
- * parameters:
- * - MessageIterator &i, an iterator over the message being parsed
- * - UpdateSchemeList &aList, a pointer to an object into which to put the result
- * returns: the updated pointer p pointing after the parsed list
+
+/**
+ * Deserialize the object from the given MessageIterator into provided UpdateSchemeList
+ *
+ * @param[in] i An iterator over the message being parsed.
+ * @param[in] aList A reference to an object to contain the result
+ *
+ * @retval WEAVE_NO_ERROR  On success.
+ * @retval WEAVE_ERROR_BUFFER_TOO_SMALL Message was too short.
+ * @retval WEAVE_ERROR_INVALID_LIST_LENGTH If the message contained an invalid list length (either not enough data to fill in the
+ * list or too many to fit within the limits)
  */
-WEAVE_ERROR UpdateSchemeList::parse(MessageIterator &i, UpdateSchemeList &aList)
+WEAVE_ERROR UpdateSchemeList::parse(MessageIterator & i, UpdateSchemeList & aList)
 {
-  i.readByte(&aList.theLength);
-  if (!i.hasData(aList.theLength)) return WEAVE_ERROR_INVALID_LIST_LENGTH;
-  for (int j=0; j<aList.theLength; j++) i.readByte(&aList.theList[j]);
-  return WEAVE_NO_ERROR;
+    TRY(i.readByte(&aList.theLength));
+    if (!i.hasData(aList.theLength))
```

Comment:
Assertion?

Reply:
I've opted to do a return error code in the `init()` functions rather than an assertion.  I'm happy to refactor subsequently, but at the moment, I'm keeping this as is.